pkg/reconciler/ingress/resources/gateway.go

```diff
@@ -445,3 +469,79 @@ func UpdateGateway(gateway *v1beta1.Gateway, want []*istiov1beta1.Server, existi
 func isPlaceHolderServer(server *istiov1beta1.Server) bool {
 	return cmp.Equal(server, &placeholderServer, protocmp.Transform())
 }
+
+// QualifiedGatewayNamesFromContext get gateway names from context.
+func QualifiedGatewayNamesFromContext(ctx context.Context, obj kmeta.Accessor) (map[v1alpha1.IngressVisibility]sets.Set[string], error) {
+	ret := make(map[v1alpha1.IngressVisibility]sets.Set[string])
+
+	gateways, err := GatewaysFromContext(ctx, obj)
+	if err != nil {
+		return ret, fmt.Errorf("failed to get gateways from configuration: %w", err)
+	}
+
+	for _, visibility := range []v1alpha1.IngressVisibility{v1alpha1.IngressVisibilityClusterLocal, v1alpha1.IngressVisibilityExternalIP} {
+		ret[visibility] = sets.New[string]()
+
+		for _, gtw := range gateways[visibility] {
+			ret[visibility] = ret[visibility].Insert(gtw.QualifiedName())
+		}
+	}
+
+	return ret, nil
+}
+
+// GatewaysFromContext get gateways relevant to this ingress from context.
+func GatewaysFromContext(ctx context.Context, obj kmeta.Accessor) (map[v1alpha1.IngressVisibility][]config.Gateway, error) {
+	ret := make(map[v1alpha1.IngressVisibility][]config.Gateway)
+
+	istioConfig := config.FromContext(ctx).Istio
+
+	// External gateways selection
+	externalGateways, err := filterGateway(istioConfig.IngressGateways, obj.GetLabels())
+	if err != nil {
+		return ret, fmt.Errorf("failed to filter external gateways: %w", err)
+	}
+
+	if len(externalGateways) == 0 {
+		externalGateways = istioConfig.DefaultExternalGateways()
+	}
+
+	ret[v1alpha1.IngressVisibilityExternalIP] = externalGateways
+
+	// Local gateways selection
+	localGateways, err := filterGateway(istioConfig.LocalGateways, obj.GetLabels())
+	if err != nil {
+		return ret, fmt.Errorf("failed to filter local gateways: %w", err)
+	}
+
+	if len(localGateways) == 0 {
+		localGateways = istioConfig.DefaultLocalGateways()
+	}
+
+	ret[v1alpha1.IngressVisibilityClusterLocal] = localGateways
+
+	return ret, nil
+}
+
+func filterGateway(gtws []config.Gateway, ingressLabels map[string]string) ([]config.Gateway, error) {
+	ret := make([]config.Gateway, 0, 1)
+
+	for _, gtw := range gtws {
+		if gtw.LabelSelector == nil { // default value
+			continue
+		}
+
+		selector, err := metav1.LabelSelectorAsSelector(gtw.LabelSelector)
+		if err != nil {
+			return ret, fmt.Errorf("failed to create selector from gateway (%s) label selector: %w", gtw.QualifiedName(), err)
+		}
```

Comment:
We should validate the selectors when validating the configmap so the operator can fix it immediately.